ssh/browser/browser.go

```diff
@@ -0,0 +1,213 @@
+			log.Printf("%v", w)
+		}
+	}
+}
+
+//go:embed dist/* index.html
+var embeddedFS embed.FS
+
+func runServe() {
+	log.Printf("Listening on %s", *addr)
+	err := http.ListenAndServe(*addr, http.FileServer(http.FS(embeddedFS)))
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// commonSetup performs setup that is common to both dev and build modes.
+func commonSetup() (*esbuild.BuildOptions, error) {
+	// Change cwd to to where this file lives -- that's where all inputs for
+	// esbuild and other build steps live.
+	if _, filename, _, ok := runtime.Caller(0); ok {
+		if err := os.Chdir(path.Dir(filename)); err != nil {
+			return nil, fmt.Errorf("Cannot change cwd: %v", err)
+		}
+	}
+	if err := buildDeps(); err != nil {
+		return nil, fmt.Errorf("Cannot build deps: %v", err)
+	}
+
+	return &esbuild.BuildOptions{
+		EntryPoints: []string{"src/index.js", "src/index.css"},
+		Loader:      map[string]esbuild.Loader{".wasm": esbuild.LoaderFile},
+		Outdir:      "./dist",
+		Bundle:      true,
+		Sourcemap:   esbuild.SourceMapLinked,
+		LogLevel:    esbuild.LogLevelInfo,
+		Define:      map[string]string{"DEBUG": strconv.FormatBool(*dev)},
+		Target:      esbuild.ES2017,
+	}, nil
+}
+
+// buildDeps builds the static assets that are needed for the server (except for
+// JS/CSS bundling, which is  handled by esbuild).
+func buildDeps() error {
+	if err := copyWasmExec(); err != nil {
+		return fmt.Errorf("Cannot copy wasm_exec.js: %v", err)
+	}
+	if err := buildWasm(); err != nil {
+		return fmt.Errorf("Cannot build main.wasm: %v", err)
+	}
+	if err := installJsDeps(); err != nil {
+		return fmt.Errorf("Cannot install JS deps: %v", err)
+	}
+	return nil
+}
+
+// copyWasmExec grabs the current wasm_exec.js runtime helper library from the
+// Go toolchain.
+func copyWasmExec() error {
+	log.Printf("Copying wasm_exec.js...\n")
+	wasmExecSrcPath := filepath.Join(runtime.GOROOT(), "misc", "wasm", "wasm_exec.js")
+	wasmExecDstPath := filepath.Join("src", "wasm_exec.js")
+	contents, err := os.ReadFile(wasmExecSrcPath)
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(wasmExecDstPath, contents, 0600)
+}
+
+// buildWasm builds the Tailscale wasm binary and places it where the JS can
+// load it.
+func buildWasm() error {
+	log.Printf("Building wasm...\n")
+	args := []string{"build", "-tags", "tailscale_go,osusergo,netgo,nethttpomithttp2,omitidna,omitpemdecrypt"}
+	if !*dev {
+		// Omit long paths and debug symbols in release builds, to reduce the
+		// generated WASM binary size.
+		args = append(args, "-trimpath", "-ldflags", "-s -w")
+	}
+	args = append(args, "-o", "src/main.wasm", "./wasm")
+	cmd := exec.Command(filepath.Join(runtime.GOROOT(), "bin", "go"), args...)
+	cmd.Env = append(os.Environ(), "GOOS=js", "GOARCH=wasm")
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+// installJsDeps installs the JavaScript dependencies specified by package.json
+func installJsDeps() error {
```

Comment:
installJSDeps per https://github.com/golang/go/wiki/CodeReviewComments#initialisms but given as how this isn't exported API I don't care so much.